enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+       - 100:1::1/64
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: TCP
+        tcp:
+          ports: "100-200"
+      action: Allow
+    - order: 20
+      protocolConfig:
+        protocol: UDP
+        udp:
+          ports: 8000
+      action: Allow
+  - sourceCIDRs:
+       - 3.3.3.3/24
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: ICMP
+        icmp:
+          icmpType: 3 # Destination unreachable
+          icmpCode: 1 # Host unreachable
+      action: Allow
+  - sourceCIDRs:
+       - 10:10::1/64
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: ICMPv6
+        icmpv6:
+          icmpType: 128 # Ping request
+      action: Deny
+  - sourceCIDRs:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: TCP
+      action: Allow
+```
+
+In the case that there are multiple `IngressNodeFirewall` objects for the same nodes and interfaces with overlapoing CIDRs, the `order` field will specify which rule applied first, lower order will be executed first and once the rule is executed and an action is taken, the firewall operation will be completed, and all the following rules won't be checked. 
+
+- eBPF data structures and types
+
+https://github.com/openshift/ingress-node-firewall/blob/master/bpf/ingress_node_firewall.h
+
+- eBPF statistics sample
+
+```shell
+curl 127.0.0.01:39301/metrics | grep -i "ingressnodefirewall_node"
+
+ingressnodefirewall_node_packet_allow_bytes 0
+
+ingressnodefirewall_node_packet_allow_total 0
+
+ingressnodefirewall_node_packet_deny_bytes 490
+
+ingressnodefirewall_node_packet_deny_total 5
+```
+
+- eBPF drop events example
+
+```shell
+2022-08-19 16:55:08 +0000 UTC kind-worker2 ruleId 10 action Drop len 98 if eth0
+
+2022-08-19 16:55:08 +0000 UTC kind-worker2 	ipv4 src addr 172.18.0.1 dst addr 172.18.0.4
+
+2022-08-19 16:55:08 +0000 UTC kind-worker2 	icmpv4 type 8 code 0
+```
+### User Stories
+https://issues.redhat.com/browse/SDN-2781
+
+### API Extensions
+
+A new cluster scoped CRD is introduced to support this feature
+
+all the new APIs for the new CRD are defined here https://github.com/openshift/ingress-node-firewall/tree/master/api/v1alpha1
+
+### Test Plan
+
+- Unit tests coverage
+- IPv4/IPv6 E2E coverage to validate ingress node firewall functionality on OCP cluster with multiple external traffic streams to match allow and deny actions and verify statistics per rule.
+
+### Risks and Mitigations
+
+admin could leave the cluster in unusable state by configuring bad rules to block traffic to api server or drop DNS,
+ingress node firewall controller will check and prevent configuration that might impact cluster basic operation.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+N/A
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+Tech Preview: 4.12
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
```

Comment:
this is already handled we have logic that merge and purge old rules so we stay in sync wuth the configs

Reply:
I'm not necessarily taking about the rules in the maps, but the EBPF map definitions themselves, how do we plan to keep the firewall up and running which swapping out the underlying map definitions?

Reply:
why would the MAP definition change ? the maps as match up with libebpf in kernel and I don't see this will ever change but if it does that will be require removing and repopulating all the maps again